File: 2019/day_01.js
// Day 1: The Tyranny of the Rocket Equation

import fs from "node:fs";

const calculateFuel = (mass) => Math.floor(mass / 3) - 2;

const calculateTotalFuel = (mass) => {
  let fuel = calculateFuel(mass);
  return fuel <= 0 ? 0 : fuel + calculateTotalFuel(fuel);
};

const partOne = (masses) => {
  return masses.reduce((acc, curr) => acc + calculateFuel(curr), 0);
};

const partTwo = (masses) => {
  return masses.reduce((acc, curr) => acc + calculateTotalFuel(curr), 0);
};

const masses = fs.readFileSync("inputs/day01.txt", "utf8").trim().split("\n");
console.log("Part one: ", partOne(masses));
console.log("Part two: ", partTwo(masses));

File: 2018/day_01.py
# Day 1: Chronal Calibration


def part_one(sequence: list[int]) -> int:
    return sum(sequence)


def part_two(sequence: list[int]) -> int:
    idx = 0
    sum = 0
    cache = set()

    while True:
        sum += sequence[idx % len(sequence)]
        if sum in cache:
            break
        cache.add(sum)

        idx += 1
    return sum


if __name__ == "__main__":
    with open("inputs/day01.txt", "r") as f:
        sequence = [int(line.strip()) for line in f]

    print("Part one: ", part_one(sequence))
    print("Part two: ", part_two(sequence))
